### ISSUE TITLE
Tips on improving multi-class classification on custom dataset

### ISSUE THREAD

**Issue by AntixK** [bentrevett/pytorch-sentiment-analysis#22]
I have a custom text dataset with little more than 13k data points with 5 classes. I tried to use your multi-class classification notebook for classifying my dataset, but it doesn't seem to work. 

Here's what I have tried -
1) Played around with the learning rates, optimizers and momentum.
2) Tried increasing and decreasing the number of conv layers.
3) Tried changing GloVe word vectors of dimensions 50, 100 and 200.

How nothing seems to work and in fact, the validation loss steadily increases. I suspect some improper embedding of my dataset. Do you have any suggestions to improve? Any help is appreciated.

**Comment by bentrevett** [comment]
The validation loss increases, but does the test loss decrease to zero? If so, you are overfitting and need to increase the amount of regularization. 

Also, are all the classes balanced?

**Comment by AntixK** [comment]
Yes, the model is clearly overfitting the data. I tried increasing the dropout a bit (0.6) but didn't work either. What other regularization methods would you suggest? Batch Norm?? :/ 

The classes are well balanced, I made sure of that.

**Comment by bentrevett** [comment]
Have you tried increasing the dropout even higher? 

What type of data are you training on? What are your inputs/outputs?

**Comment by AntixK** [comment]
This is my current CNN model with Batch Norm (switching between that and dropout)
```
class CNN(nn.Module):
    def __init__(self, vocab_size, embedding_dim, n_filters, filter_sizes, output_dim, dropout):
        super().__init__()
        
        self.embedding = nn.Embedding(vocab_size, embedding_dim)
        self.conv_0 = nn.Conv2d(in_channels = 1, out_channels = n_filters, 
                                kernel_size = (filter_sizes[0], embedding_dim))
        self.bn0 = nn.BatchNorm2d(n_filters)
        self.conv_1 = nn.Conv2d(in_channels = 1, out_channels = n_filters, 
                                kernel_size = (filter_sizes[1], embedding_dim))
        self.bn1 = nn.BatchNorm2d(n_filters)
        self.conv_2 = nn.Conv2d(in_channels = 1, out_channels = n_filters, 
                                kernel_size = (filter_sizes[2], embedding_dim))
        self.bn2 = nn.BatchNorm2d(n_filters)
        self.fc = nn.Linear(len(filter_sizes) * n_filters, output_dim)
        self.dropout = nn.Dropout(dropout)
        
    def forward(self, text):
        
        text = text.permute(1, 0)
        
        embedded = self.embedding(text)
        
        embedded = embedded.unsqueeze(1)
              
        conved_0 = F.relu(self.bn0(self.conv_0(embedded)).squeeze(3))
        conved_1 = F.relu(self.bn1(self.conv_1(embedded)).squeeze(3))
        conved_2 = F.relu(self.bn2(self.conv_2(embedded)).squeeze(3))
        
        pooled_0 = F.max_pool1d(conved_0, conved_0.shape[2]).squeeze(2)
        pooled_1 = F.max_pool1d(conved_1, conved_1.shape[2]).squeeze(2)
        pooled_2 = F.max_pool1d(conved_2, conved_2.shape[2]).squeeze(2)
        
        # cat = self.dropout(torch.cat((pooled_0, pooled_1, pooled_2), dim=1))            
        return self.fc(cat)
```
My data is basically a long text with a corresponding label. I used data.TabularDataset to create the dataset and  tokenize (using SpaCy) and then built a vocab on the training dataset using Glove.6B pretrained word vectors.
I didn't do much different from the standard procedure.

**Comment by bentrevett** [comment]
This might be useful to you. Has slides on how to debug/what steps to take when building a NN system. Let me know if you manage to solve the problem!

http://josh-tobin.com/troubleshooting-deep-neural-networks.html
